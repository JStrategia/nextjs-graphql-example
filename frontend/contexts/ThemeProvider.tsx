import React, { createContext, memo, ReactNode, useMemo, useState } from 'react';
import { ThemeProvider as StyledComponentsThemeProvider } from 'styled-components';
import { ThemeMode } from '../types/styles';
import { getTheme, GlobalStyle } from '../styles';

const defaultMode = 'light';

type Props = {
  children: ReactNode;
};

function ThemeProvider({ children }: Props) {
  const [mode, setMode] = useState<ThemeMode>(defaultMode);

  const theme = useMemo(() => getTheme(mode), [mode]);

  function switchMode() {
    setMode(mode === 'light' ? 'dark' : 'light');
  }

  return (
    <ThemeContext.Provider value={{ mode, setMode, switchMode }}>
      <StyledComponentsThemeProvider theme={theme}>
        <GlobalStyle />
        {children}
      </StyledComponentsThemeProvider>
    </ThemeContext.Provider>
  );
}

const ThemeContext = createContext<{
  mode: ThemeMode;
  setMode: (mode: ThemeMode) => void;
  switchMode: () => void;
}>({ mode: defaultMode, setMode: () => void 0, switchMode: () => void 0 });

export default memo(ThemeProvider);

export { ThemeContext };
