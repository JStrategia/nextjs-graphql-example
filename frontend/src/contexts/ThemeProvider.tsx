import React, { createContext, ReactNode, useMemo, useState } from 'react';
import { ThemeProvider as StyledComponentsThemeProvider } from 'styled-components';
import { getTheme, GlobalStyle } from '../styles';
import { ThemeMode } from '../types/styles';

const defaultMode = 'dark';

type Props = {
  children: ReactNode;
  initialMode?: ThemeMode;
};

function ThemeProvider({ children, initialMode = defaultMode }: Props) {
  const [mode, setMode] = useState<ThemeMode>(initialMode);

  const theme = useMemo(() => getTheme(mode), [mode]);

  function switchMode() {
    setMode(mode === 'light' ? 'dark' : 'light');
  }

  return (
    <ThemeContext.Provider value={{ mode, setMode, switchMode }}>
      <StyledComponentsThemeProvider theme={theme}>
        <GlobalStyle />
        {children}
      </StyledComponentsThemeProvider>
    </ThemeContext.Provider>
  );
}

const ThemeContext = createContext<{
  mode: ThemeMode;
  setMode: (mode: ThemeMode) => void;
  switchMode: () => void;
}>({ mode: defaultMode, setMode: () => void 0, switchMode: () => void 0 });

export default ThemeProvider;
export { defaultMode, ThemeContext };
