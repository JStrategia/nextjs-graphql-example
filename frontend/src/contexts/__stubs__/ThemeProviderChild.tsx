import * as React from 'react';
import { useContext } from 'react';
import { ThemeContext } from 'contexts/ThemeProvider';
import { ThemeMode } from 'types/styles';

type Props = {
  readonly modeToSet: ThemeMode;
};

function ThemeProviderChild({ modeToSet }: Props) {
  const { mode, setMode, switchMode } = useContext(ThemeContext);

  return (
    <>
      <span data-testid="mode">{mode}</span>
      <button data-testid="switchModeButton" onClick={switchMode} />
      <button data-testid="setModeButton" onClick={() => setMode(modeToSet)} />
    </>
  );
}

export default ThemeProviderChild;
