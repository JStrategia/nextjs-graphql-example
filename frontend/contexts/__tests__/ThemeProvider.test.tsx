import * as React from 'react';
import { create } from 'react-test-renderer';
import { render, cleanup, fireEvent } from '@testing-library/react';
import ThemeProvider from '../ThemeProvider';
import ThemeProviderChild from './__stubs__/ThemeProviderChild';

describe('ThemeProvider', () => {
  describe('Given children', () => {
    const modeToSet = 'dark';
    const children = <ThemeProviderChild modeToSet={modeToSet} />;

    // TODO: See if this is necessary, we have multiple data-testid for some reason
    afterEach(cleanup);

    it('Should match snapshot', () => {
      const render = create(<ThemeProvider>{children}</ThemeProvider>).toJSON();

      expect(render).toMatchSnapshot();
    });

    describe('Given no initial mode', () => {
      const { getByTestId } = render(<ThemeProvider>{children}</ThemeProvider>);

      it('Should have dark as its initial mode', () => {
        const mode = getByTestId('mode');

        expect(mode).toBe('dark');
      });
    });

    describe('Given initial mode', () => {
      const initialMode = 'light';
      const { getByTestId } = render(<ThemeProvider initialMode={initialMode}>{children}</ThemeProvider>);

      it('Should have initial mode as its mode', () => {
        const mode = getByTestId('mode');

        expect(mode).toBe(initialMode);
      });

      describe('When setting theme', () => {
        const setModeButton = getByTestId('setModeButton');
        fireEvent.click(setModeButton);

        it('Should set theme', () => {
          const mode = getByTestId('mode');

          expect(mode).toBe(modeToSet);
        });
      });

      describe('When switching theme', () => {
        const modeToBeSwitched = 'dark';

        const switchModeButton = getByTestId('switchModeButton');
        fireEvent.click(switchModeButton);

        it('Should switch theme', () => {
          const mode = getByTestId('mode');

          expect(mode).toBe(modeToBeSwitched);
        });
      });
    });
  });
});
