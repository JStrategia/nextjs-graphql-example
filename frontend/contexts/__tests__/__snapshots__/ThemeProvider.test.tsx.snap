// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThemeProvider Given children Should match snapshot 1`] = `
<span>
  Children
</span>
`;
